# Instalação das dependências

#### pip install scipy
#### pip install numpy
#### pip install scikit-learn
#### pip install jupyter
#### pip install pandas
#### pip install pip install seaborn
----
### Extra
#### pip install matplotlib


# Iris Dataset
### Importação dos dados

In [1]:
from sklearn.datasets import load_iris

iris = load_iris()

### Observações
O objeto é um numpy array, que é mais eficiente e rápido que uma lista normal do python
Sempre se usa uma variável 'x' para as observações

In [2]:
x = iris.data
print(iris.data)

[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.6 1.4 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

#### Shape das observações
Permite verificar o tamanho dos arrays

In [3]:
print(iris.data.shape)

(150, 4)


## Target
### Mostra os nomes das espécies
print(iris.target_names)
#### Mostras os valores das espécies
print(iris.target)
precisa ter os dados sempre com um array numpy para as observações e outro para o target os dois devem ser somente numéricos
O target fica em uma variável 'y'


In [4]:
y = iris.target
print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


Temos um array com 150 respostas e 4 atributos que serão chamadas de features

#### Shape do Target
Permite verificar o tamanho dos arrays

In [5]:
print(iris.target.shape)

(150,)


Temos um array com 150 obervações, esse número precisa fechar com o número de respostas

### Testando com o KNN
Será necessário definir um k

### Importação do KNN

In [64]:
from sklearn.neighbors import KNeighborsClassifier
# n_neighbors=1 valor de k
knn = KNeighborsClassifier(n_neighbors=3)

### Treinar o modelo
Lembrando que 
x = data
y = target

In [65]:
knn.fit(x,y)

KNeighborsClassifier(n_neighbors=3)

### Realizando uma previsão
knn.predict([observações])

In [66]:
knn.predict([[5.1, 3.5, 1.4, 0.2]])

array([0])

In [67]:
#O [0] no final indica que queremos pegar apenas o primeiro índice do vetor resultante
especies = knn.predict([[5.9, 3,  5.1, 1.8]])[0]
#Baseado na especie retornada, vamos obter a espécie do array target_names
print(iris.target_names[especies])

virginica


### Separar os dados em dois grupos

In [68]:
from sklearn.model_selection import train_test_split
#divide o conjunto em teste e treino o 0.5 siginifca 50% para cada, se usar 0.25 significa que 75% é para treinamento e 25% para teste
#a própria função se encarrega de sortear os dados
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.25)

### Verificando o tamanho dos conjuntos de treinamento e teste

In [69]:
print(x_train.shape)
print(x_test.shape)

(112, 4)
(38, 4)


### Avaliando o modelo

In [70]:
knn.fit(x_train,y_train)
previsoes = knn.predict(x_test)
print(previsoes)

[1 2 1 0 1 1 2 0 1 2 2 2 1 1 0 0 1 0 2 2 1 0 1 2 1 1 0 2 0 0 0 0 0 1 2 0 2
 0]


In [71]:
#comparando se os resultados foram corretos
from sklearn import metrics
acertos = metrics.accuracy_score(y_test, previsoes)
print(acertos)

0.9736842105263158


### Aplicando regressão logística

In [72]:
from sklearn.linear_model import LogisticRegression
# multi_class="ovr" significa que se está usando o métodos one vs all para determinar a classe, já que regressão logística é binária. No entando está é a configuração padrão.
logreg = LogisticRegression()
logreg.fit(x_train, y_train)
#Permite listar a probabilidade de cada classe
print(logreg.predict_proba([[6.2, 3.4, 5.4, 2.3]]))
#imprimindo a classe prevista
print(logreg.predict([[6.2, 3.4, 5.4, 2.3]]))
previsoes_logreg = logreg.predict(x_test)
acertos_logreg = metrics.accuracy_score(y_test, previsoes_logreg)
print(acertos_logreg)

[[9.49356935e-05 4.81444506e-02 9.51760614e-01]]
[2]
0.9736842105263158
